src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -438,22 +480,36 @@ internal static ODataUriResolver GetUriResolver(IServiceProvider container)
             return container.GetRequiredService<ODataUriResolver>();
         }
 
-        private static List<T> FindAcrossModels<T>(IEdmModel model, String qualifiedName, bool caseInsensitive) where T : IEdmSchemaElement
+        private static IList<T> FindAcrossModels<T>(IEdmModel model, String qualifiedName, bool caseInsensitive) where T : IEdmSchemaElement
         {
-            List<T> results = FindSchemaElements<T>(model, qualifiedName, caseInsensitive).ToList();
+            IList<T> results = null;
+            FindSchemaElements<T>(model, qualifiedName, caseInsensitive, ref results);
 
             foreach (IEdmModel reference in model.ReferencedModels)
             {
-                results.AddRange(FindSchemaElements<T>(reference, qualifiedName, caseInsensitive));
+                FindSchemaElements<T>(reference, qualifiedName, caseInsensitive, ref results);
             }
 
             return results;
         }
 
-        private static IEnumerable<T> FindSchemaElements<T>(IEdmModel model, string qualifiedName, bool caseInsensitive) where T : IEdmSchemaElement
+        private static void FindSchemaElements<T>(IEdmModel model, string qualifiedName, bool caseInsensitive, ref IList<T> results) where T : IEdmSchemaElement
         {
-            return model.SchemaElements.OfType<T>()
-            .Where(e => string.Equals(qualifiedName, e.FullName(), caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal));
+            foreach (var schema in model.SchemaElements)
+            {
+                if (string.Equals(qualifiedName, schema.FullName(), caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))
+                {
+                    if (schema is T)
+                    {
+                        if (results == null)
+                        {
+                            results = new List<T>();
+                        }
+
+                        results.Add((T)schema);

```

Comment:
>(( [](start = 35, length = 2)

Try and get rid of this repeated cast.

Reply:
Does not help. If it is a class, we can but not with an interface. 

---
In reply to: [370899362](https://github.com/OData/odata.net/pull/1602#discussion_r370899362) [](ancestors = 370899362)